### PR TITLE
Add `host` option

### LINKED ISSUE #3: `host` option
```js
> require("net").createServer().listen("1234", "127.0.0.1")
> require("net").createServer().listen("1234", "127.0.0.2")
```

```
$ ss | grep 1234
LISTEN   0        0                127.0.0.2:1234             0.0.0.0:*       users:(("node",pid=1561254,fd=22))
LISTEN   0        0                127.0.0.1:1234             0.0.0.0:*       users:(("node",pid=1561255,fd=21))
```

A `host` option is needed to distinguish. Could accept the same signatures as [net.Server#listen](https://nodejs.org/api/net.html#net_server_listen_port_host_backlog_callback).

### PR DESCRIPTION
**Pull request by nagygergo**
Fixes #3. 

TODO: Update `.all()`

How should the `.all()` method be supported? 
My proposal would be that `.all()` returns a `{port}-{pid}` map for all the hosts and `.all(host)` `{port}-{pid}` map for the given host.

TODO: Update Readme with examples

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -61,18 +62,33 @@ const getList = async () => {
 		.map(line => line.match(/\S+/g) || []);
 };
 
-module.exports.portToPid = async port => {
-	if (Array.isArray(port)) {
+module.exports.portToPid = async (arg1, arg2) => {
```
Comment (sindresorhus) [inline index.js:65]: `host` should be an option in an options-object.

File: test.js [hunk 2]
```diff
@@ -43,3 +43,19 @@ test('`.list()`', async t => {
 	t.true(all instanceof Map);
 	await t.notThrowsAsync(pidPort.portToPid([...all.keys()]));
 });
+
+test('Node `server.listen()` signature - with different host as arg', async t => {
```
Comment (sindresorhus) [inline test.js:47]: Don't use abbreviations.

File: index.js [hunk 3]
```diff
@@ -61,18 +62,33 @@ const getList = async () => {
...
+	let host;
+	let port;
+	if (typeof arg1 === 'object' && arg1.port) {
+		port = arg1.port;
+		host = arg1.host;
+	} else if (typeof arg2 === 'string') {
+		port = arg1;
```
Comment (sindresorhus) [inline index.js:72]: Don't use abbreviations.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
> My proposal would be that .all() returns a {port}-{pid} map for all the hosts and .all(host) {port}-{pid} map for the given host.

I think it needs to include the host too. So maybe: `Promise<Map<number, {host: number, port: number}>>`. And yes, it should have a `host` option.

File: index.js [hunk 4]
```diff
@@ -41,8 +41,9 @@ const parsePid = pid => {
 	}
 };
 
-const getPort = (port, list) => {
-	const regex = new RegExp(`[.:]${port}$`);
+const getPort = (port, list, host) => {
```
Comment (sindresorhus) [inline index.js:44]:
```suggestion
const getPort = (host, port, list) => {
```

**Comment by sindresorhus** [comment]
And `host` should have a default.

**Comment by nagygergo** [comment]
> > My proposal would be that .all() returns a {port}-{pid} map for all the hosts and .all(host) {port}-{pid} map for the given host.
> 
> I think it needs to include the host too. So maybe: `Promise<Map<number, {host: number, port: number}>>`. And yes, it should have a `host` option.

I don't think `Promise<Map<number, {host: number, port: number}>>` would work, becuase a single pid could have multiple ports associated to it. So I think the `Promise<Map<number,number>` with pid-port map is ok.

File: index.js (hunk 1 above)
Comment (nagygergo) [reply index.js:65]: The `net.server.listen(port, host)` signature is valid, but leaving that out is fine for me.

**Comment by nagygergo** [comment]
> And `host` should have a default.

My understanding is that if the host is not specified the mapping between the port and the pid  should be done for all the local addresses that executing machine owns, like 127.0.0.1, 192.168.8.8, maybe a non-local subnet IP.

I don't know what would be a good default for scanning.

**Comment by sindresorhus** [comment]
> I don't think Promise<Map<number, {host: number, port: number}>> would work, becuase a single pid could have multiple ports associated to it. So I think the Promise<Map<number,number> with pid-port map is ok.

Just a typo. I meant: `Promise<Map<number, {host: string, pid: number}>>`

**Comment by sindresorhus** [comment]
> My understanding is that if the host is not specified the mapping between the port and the pid should be done for all the local addresses that executing machine owns, like 127.0.0.1, 192.168.8.8, maybe a non-local subnet IP.
> I don't know what would be a good default for scanning.

That makes sense to me. The behavior needs to be documented though.

@silverwind Thoughts on this?

**Comment by sindresorhus** [comment]
// @Richienb In case you're interested in helping reviewing this.

**Comment by nagygergo** [comment]
Hey @Richienb, @sindresorhus. 

Just going through some edge-cases for the host-port-pid mapping, the issue is that it's a many-to-many-to-many mapping. The current api signatures do not really cater for these cases. So in the following cases the current methods have unspecified behavior:

1. A machine is using multiple local addresses , but has a processes occupying the same port (127.0.0.1:1212, 127.0.0.2:1212).  (this can easily happen when virtualization agent is creating virtual networks for it's VMs, like KVM or Minikube)
2. A machine is running a cluster of workers to process the same port, in this case there are many pids associated to the same host:port combination (eg. httpd)
```
Netid  State      Recv-Q Send-Q Local Address:Port     Peer Address:Port
tcp    LISTEN     0      128    :::80                :::*                 users:(("httpd",pid=10522,fd=4),("httpd",pid=10521,fd=4),("httpd",pid=10520,fd=4),("httpd",pid=10519,fd=4),("httpd",pid=10518,fd=4),("httpd",pid=10516,fd=4))
```
3. A process is consuming more than 1 port or more than 1 port across different local addresses. (I don't know a good example to this)

Not sure about the exact solution yet, will think about it next week, and I'm not sure if there is a need to handle these kinds of usecases, but I wanted to point them out anyways.

**Comment by silverwind** [comment]
> My understanding is that if the host is not specified the mapping between the port and the pid should be done for all the local addresses that executing machine owns, like 127.0.0.1, 192.168.8.8, maybe a non-local subnet IP.
I don't know what would be a good default for scanning.

From [node docs](https://nodejs.org/api/net.html#net_server_listen_port_host_backlog_callback):

> If host is omitted, the server will accept connections on the unspecified IPv6 address (::) when IPv6 is available, or the unspecified IPv4 address (0.0.0.0) otherwise.

Unspecified IPv6 address means dual-stack behaviour so it will match all local IPv4 and IPv6 interface when IPv6 is available (not sure how to check), so I think it's fine for this module to return all PIDs that listen on the specified port, IPv4 and IPv6.

**Comment by sindresorhus** [comment]
@nagygergo Still interested in finishing this PR? 🙏

**Comment by sindresorhus** [comment]
Bump :)

**Pull request closed without merging**